### PR TITLE
perf(animations): made errors in the animations package tree shakeable

### PR DESCRIPTION
**Pull request by jessicajaniuk**
This moves all the error strings into exported functions that can be tree shaken away.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Errors in the animations code base are just defined all inline and affect bundle size.

Issue Number: N/A

## What is the new behavior?
Errors are centralized in one file and should be tree shakeable.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/animations/browser/src/errors.ts [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+const LINE_START = '\n - ';
+
+export function invalidTimingValue(exp: string|number): Error {
+  return new Error(`The provided timing value "${exp}" is invalid.`);
```
Comment (AndrewKushnir) [inline packages/animations/browser/src/errors.ts:12]: I think it'd be great to use `RuntimeError` class from the `@angular/core`, so it's more consistent with other packages.

We'd also need to use `ngDevMode` to make error messages (strings) tree-shakable.

```suggestion
  return new RuntimeError(
    RuntimeErrorCode.INVALID_TIMIMG_VALUE,
    NG_DEV_MODE && `The provided timing value "${exp}" is invalid.`);
```
where `NG_DEV_MODE` can be declared at the top of the file as:

```
const NG_DEV_MODE = typeof ngDevMode === 'undefined' || !!ngDevMode;
```

Without the `NG_DEV_MODE`, the build optimizer wouldn't be able to tree-shake the string away.

Comment (jessicajaniuk) [reply packages/animations/browser/src/errors.ts:12]: I initially was going to do that, but then saw the reactive forms errors and based it off of that instead. I'll switch over. Thanks for the heads up.

File: goldens/size-tracking/integration-payloads.json [hunk 2]
```diff
@@ -60,8 +60,8 @@
   "animations": {
     "master": {
       "uncompressed": {
-        "runtime": 1073,
-        "main": 172602,
+        "runtime": 1070,
+        "main": 171374,
```
Comment (AndrewKushnir) [inline goldens/size-tracking/integration-payloads.json:64]: 🎉

File: packages/animations/browser/src/dsl/animation_transition_expr.ts [hunk 3]
```diff
@@ -34,7 +37,8 @@ function parseInnerTransitionStr(
 
   const match = eventStr.match(/^(\*|[-\w]+)\s*(<?[=-]>)\s*(\*|[-\w]+)$/);
   if (match == null || match.length < 4) {
-    errors.push(`The provided transition expression "${eventStr}" is not supported`);
+    // TODO(jessicajaniuk): use error method from errors.ts
```
Comment (AndrewKushnir) [inline packages/animations/browser/src/dsl/animation_transition_expr.ts:40]: Drop TODO (it looks like it's done)?

File: packages/animations/browser/src/dsl/animation_transition_expr.ts [hunk 4]
```diff
@@ -60,7 +64,8 @@ function parseAnimationAlias(alias: string, errors: string[]): string|Transition
     case ':decrement':
       return (fromState: any, toState: any): boolean => parseFloat(toState) < parseFloat(fromState);
     default:
-      errors.push(`The transition alias value "${alias}" is not supported`);
+      // TODO(jessicajaniuk): use error method from errors.ts
```
Comment (AndrewKushnir) [inline packages/animations/browser/src/dsl/animation_transition_expr.ts:67]: Drop TODO (it looks like it's done)?

File: packages/animations/browser/src/errors.ts [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+  ANIMATION_FAILED = 3502,
+  REGISTRATION_FAILED = 3503,
+  CREATE_ANIMATION_FAILED = 3504,
+  TRANSITION_FAILED = 3505
+}
+
+export function invalidTimingValue(exp: string|number): Error {
+  return new RuntimeError(
+      RuntimeErrorCode.INVALID_TIMING_VALUE,
+      NG_DEV_MODE ? `The provided timing value "${exp}" is invalid.` : '');
```
Comment (AndrewKushnir) [inline packages/animations/browser/src/errors.ts:61]: nit: I think you should be able to use a bit more compact version (the `RuntimeError` class understands `false` as a second argument):
```suggestion
      NG_DEV_MODE && `The provided timing value "${exp}" is invalid.`);
```

File: packages/animations/browser/src/render/animation_engine_next.ts [hunk 6]
```diff
@@ -42,12 +43,12 @@ export class AnimationEngine {
...
     let trigger = this._triggerCache[cacheKey];
     if (!trigger) {
-      const errors: string[] = [];
+      const errors: Error[] = [];
       const ast =
           buildAnimationAst(this._driver, metadata as AnimationMetadata, errors) as TriggerAst;
       if (errors.length) {
-        throw new Error(`The animation trigger "${
-            name}" has failed to build due to the following errors:\n - ${errors.join('\n - ')}`);
+        // TODO(jessicajaniuk): use error method from errors.ts
```
Comment (AndrewKushnir) [inline packages/animations/browser/src/render/animation_engine_next.ts:50]: Drop TODO?

File: packages/animations/browser/src/render/shared.ts [hunk 7]
```diff
@@ -81,6 +81,7 @@ export function normalizeKeyframes(
   });
   if (errors.length) {
     const LINE_START = '\n - ';
+    // TODO(jessicajaniuk): use error method from errors.ts
     throw new Error(
```
Comment (AndrewKushnir) [inline packages/animations/browser/src/render/shared.ts:85]: Would that be straightforward to include this one into the `errors.ts` too?

**Review by AndrewKushnir — COMMENTED** [review]
LGTM, just some minor comments.

I think we'd need to add golden files for the error codes as well, see https://github.com/angular/angular/pull/44677, so that they are a part of the public API.

File: packages/animations/browser/src/dsl/animation_transition_expr.ts (hunk 3 above)
Comment (jessicajaniuk) [reply packages/animations/browser/src/dsl/animation_transition_expr.ts:40]: Good catch. I was clearing these out as I went and forgot a couple.

File: packages/animations/browser/src/render/shared.ts (hunk 7 above)
Comment (jessicajaniuk) [reply packages/animations/browser/src/render/shared.ts:85]: Yes. I prepared the error for it and then missed it when applying all the methods. Easy fix.

File: packages/animations/browser/src/errors.ts (hunk 5 above)
Comment (jessicajaniuk) [reply packages/animations/browser/src/errors.ts:61]: Awesome. Easy fix.

**Review by atscott — APPROVED** [review]
reviewed-for: size-tracking

File: goldens/public-api/animations/errors.md [hunk 8]
```diff
@@ -0,0 +1,193 @@
+## API Report File for "angular-srcs"
+
+> Do not edit this file. It is a report generated by [API Extractor](https://api-extractor.com/).
+
+```ts
+
+// @public (undocumented)
+export function animationFailed(errors: Error[]): Error;
```
Comment (alxhub) [inline goldens/public-api/animations/errors.md:8]: Is there value in tracking these signatures from a public API perspective?

In other "error" API tests that we have, we only really care about the error codes, not the helper functions that create those errors.

Comment (jessicajaniuk) [reply goldens/public-api/animations/errors.md:8]: I think I could move the helper functions and only have the codes exposed.

Comment (jessicajaniuk) [reply goldens/public-api/animations/errors.md]: They are now moved.

**Review by alxhub — APPROVED** [review]
Reviewed-for: public-api

**Review by atscott — APPROVED** [review]
reviewed-for: public-api

**Review by AndrewKushnir — APPROVED** [review]
Reviewed-for: public-api

**Review by AndrewKushnir — APPROVED** [review]
Reviewed-for: size-tracking

**Comment by alxhub** [comment]
This PR doesn't apply cleanly to 13.2.x. Marking it as `master`-only - please open a separate PR against the patch branch if desired.

**Comment by alxhub** [comment]
This PR was merged into the repository by commit 4c778cdb28de128c6ddecd2fd6cb6257d675500b.

**Pull request closed without merging**